### PR TITLE
Add episode 11

### PR DESCRIPTION
**Pull request by bitsondatadev**
Don't merge until Monday.

@sopel39 can you take a look and let me know if anything needs to be changed? Then if all looks well, could you merge on Monday the 22nd?

### CODE REVIEW

**Comment by sopel39** [comment]
@raunaqmorarka please also take a look

File: _episodes/11.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+tables that when joined provide human readable context to enrich the recordings
+in the fact table. The schema ends up looking like a star with the fact table at
+the center. In essence, you just need to remember when someone is describing a
+fact table they are saying it is a bigger table that is likely going to end up
+on the probe side of a join, where a dimension is more likely a candidate to fit
+into memory on the build side of a join.
+
+So let's get onto the Dynamic Filtering shall we?
+
+Dynamic Filtering takes advantage of joins with big fact tables to smaller
```
Comment (sopel39) [inline _episodes/11.md:117]: Make it specific that it's node local dynamic filtering?

File: _episodes/11.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+answer the query. We can look at the blog written for the original local dynamic
+filtering implementation by Roman Zeyde.
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+One aspect that could make this better is if we have a probe side that is 
+partitioned, and a predicate exists
+
+
+You can require that these predicates exist in queries if you want to enforce
```
Comment (sopel39) [inline _episodes/11.md:134]: I don't understand this sentence.

File: _episodes/11.md (hunk 1 above)
Comment (bitsondatadev) [reply _episodes/11.md:117]: Can do!

Comment (bitsondatadev) [reply _episodes/11.md:117]: So actually my intent was to talk about the generic premise of dynamic filters and the last sentence I allude to looking at the local dynamic filtering blog for more understanding on how that works. Then talk a bit about how local is different from dynamic partition pruning.

File: _episodes/11.md (hunk 2 above)
Comment (bitsondatadev) [reply _episodes/11.md]: whoops. that's because I didn't finish it :) I revised this whole section and added what I meant to put here. Sorry about that. Please reread when you have the chance.

File: _episodes/11.md (hunk 1 above)
Comment (bitsondatadev) [reply _episodes/11.md:118]: Updated this

File: _episodes/11.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+execution? In the next iteration of Dynamic Filtering, called Dynamic Partition 
+Pruning, we can take advantage of probe tables that are partitioned and use that
+information to tell the coordinator to prune or eliminate the splits that do not
+need to be scheduled. This only works, however, if the query contains a
+predicate on the column that the partition exists on. You can require that these
+predicates exist in all queries. If you want to enforce this for dynamic
+filtering by setting this property.
+
+{% highlight text %}hive.query-partition-filter-required=true
+{% endhighlight %}
```
Comment (raunaqmorarka) [inline _episodes/11.md:144]: This part is unrelated to dynamic filtering. `hive.query-partition-filter-required` would require all queries which scan partitioned table to explicitly contain a static filter on it.

File: _episodes/11.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+With that basic understanding, let's move on to the PR that implemented Dynamic
+Partition Pruning!
+
+## PR of the week: PR 1072 Implement Dynamic Partition Pruning
+
+In this week's pull request <https://github.com/trinodb/trino/pull/1072> we
+return with Raunaq Morarka and Karol Sobczak. This PR effectively brings in the 
+second iteration of Dynamic Filtering, dynamic partition pruning, where instead
+of relying on local dynamic filtering we involved the coordinator to update the
+query plan and prune out extra splits that wouldn't be needed based on the
```
Comment (raunaqmorarka) [inline _episodes/11.md:155]: we don't update the query plan in coordinator, we collect dynamic filters from the workers there and use them when they're ready

Comment (bitsondatadev) [reply _episodes/11.md:155]: ah right, I forgot to change this..you corrected that during the show. Fixing.

File: _episodes/11.md (hunk 3 above)
Comment (bitsondatadev) [reply _episodes/11.md:144]: removing!

File: _episodes/11.md (hunk 4 above)
Comment (bitsondatadev) [reply _episodes/11.md]: Fixed

File: _episodes/11.md (hunk 3 above)
Comment (bitsondatadev) [reply _episodes/11.md]: Removed

File: _episodes/11.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+implementation for Dynamic Filtering before Raunaq's changes.
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+Local Dynamic Filtering is definitely beneficial, in that it generates a list
+of values that can be used to avoid sending some unnecessary rows, but what if 
+we could cut out entire splits of unnecessary partitions from the query
+execution? In the next iteration of Dynamic Filtering, called Dynamic Partition 
+Pruning, we can take advantage of probe tables that are partitioned and use that
+information to tell the coordinator to prune or eliminate the splits that do not
```
Comment (raunaqmorarka) [inline _episodes/11.md:127]: ... tell the coordinator to avoid generating splits from partitions that can be eliminated based on the dynamic filters.

Maybe split up this sentence

File: _episodes/11.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+need to be scheduled. This only works, however, if the query contains a
+predicate on the column that the partition exists on. You can require that these
```
Comment (raunaqmorarka) [inline _episodes/11.md:129]: ...contains a join condition ...

File: _episodes/11.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+predicates exist in all queries.
```
Comment (raunaqmorarka) [inline _episodes/11.md:130]: You can require that these predicates exist in all queries. 

Remove above line

File: _episodes/11.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+date: 2021-02-18
+tags: query planning cost optimizer dynamic filter
+youtube_id: "-2SjSM73l8g"
+wistia_id: "mj47uifu98"
+sections: 
+   - title: "Concept of the week"
+     desc: "Dynamic Filtering and Recap"
+     time: 463
+   - title: "PR of the week"
+     desc: "PR 1072 Implement Dynamic Partition Pruning"
```
Comment (mosabua) [inline _episodes/11.md:13]: Implement dynamic partition pruning

File: _episodes/11.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+     time: 1076
+   - title: "PR Demo"
+     desc: "PR Demo: PR 1072 Implement Dynamic Partition Pruning"
```
Comment (mosabua) [inline _episodes/11.md:16]: PR demo: PR 1071 Implement dynamic partition pruning

File: _episodes/11.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+     time: 2523
+---
+
+Trino nation, We want to hear from you! If you have a question or pull request 
+that you would like us to feature on the show please join the 
+[Trino slack](slack.html) and go to the \#trino-community-broadcast channel
```
Comment (mosabua) [inline _episodes/11.md:22]: , and go

File: _episodes/11.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+that you would like us to feature on the show please join the 
+[Trino slack](slack.html) and go to the \#trino-community-broadcast channel 
+and let us know there. Otherwise, you can message Manfred Moser or Brian Olsen 
+directly. Also, feel free to reach out to us on our Twitter channels Brian 
+[@bitsondatadev](https://twitter.com/bitsondatadev) and Manfred 
+[@simpligility](https://twitter.com/simpligility).
+
+If you want to show us some 💕, please [give us a ⭐ on Github](https://github.com/trinodb/trino/blob/master/.github/star.png).
+
+## Release 352
```
Comment (mosabua) [inline _episodes/11.md:30]: Empty line after title

File: _episodes/11.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+Release Notes discussed: <https://trino.io/docs/current/release/release-352.html>
```
Comment (mosabua) [inline _episodes/11.md:31]: Release notes

File: _episodes/11.md [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+
+No new release to discuss yet except that 353 will be around the corner to fix
+a low-impact correctness issue that came out in 352 
+<https://github.com/trinodb/trino/pull/6895>.
+
+## Concept of the week: Dynamic Filtering
```
Comment (mosabua) [inline _episodes/11.md:37]: Empty line after title

File: _episodes/11.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+So we've covered a lot on the Trino Community Broadcast to build our way up
+to tackling this pretty big subject in the space called Dynamic Filtering. If
+you haven't seen episodes 5-9, you may want to go back and watch those for some
+context for this episode. Episode 8 actually diverted to the Trino rebrand so
+we won't discuss that one. For the recap;
+
+In [episode 5](/broadcast/episodes/5.html), we spoke about hive partitions. In
```
Comment (mosabua) [inline _episodes/11.md:44]: Hive

File: _episodes/11.md [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+## Concept of the week: Dynamic Filtering
+So we've covered a lot on the Trino Community Broadcast to build our way up
+to tackling this pretty big subject in the space called Dynamic Filtering. If
+you haven't seen episodes 5-9, you may want to go back and watch those for some
+context for this episode. Episode 8 actually diverted to the Trino rebrand so
+we won't discuss that one. For the recap;
+
+In [episode 5](/broadcast/episodes/5.html), we spoke about hive partitions. In 
+order to save you time when you run a query, Hive stores data under directories
+named by the values of the data being written underneath that directory. take
```
Comment (mosabua) [inline _episodes/11.md:46]: Take

File: _episodes/11.md [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+this directory structure for the `orders` table partitioned by the `orderdate`
+field.
```
Comment (mosabua) [inline _episodes/11.md:48]: field:

File: _episodes/11.md [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+│   └── ...
+├── orderdate=1992-01-02
+│   └── ...
+├── orderdate=1992-01-03
+│   └── ...
+└── ...
+{% endhighlight %}
+
+When querying for data under January 1st, 1992, according to the Hive model,
+query engines like Hive and Trino will only scan orc files under the
```
Comment (mosabua) [inline _episodes/11.md:65]: only scan ORC files under

File: _episodes/11.md [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+│   └── ...
+└── ...
+{% endhighlight %}
+
+When querying for data under January 1st, 1992, according to the Hive model,
+query engines like Hive and Trino will only scan orc files under the 
+`orders/orderdate=1992-01-01` directory. The idea is to avoid scanning
+unnecessary data by grouping rows based on a field commonly used in a query.
+
+In episode 6/7, we discussed a bit about how a query gets represented internally
```
Comment (mosabua) [inline _episodes/11.md:69]: 6 and 7 .. 

or even better

six and seven

File: _episodes/11.md [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+to Trino once you submit your SQL query. First, the Parser will convert SQL to
```
Comment (mosabua) [inline _episodes/11.md:70]: parser converts SQL

File: _episodes/11.md [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+an abstract syntax tree (AST) format. Then the Planner generates a different
+ tree structure called the intermediate represenation (IR) that contains nodes 
+representing the steps that need to be performed in order to answer the query. 
+The leaves of the tree get executed first, and the parents of each node are 
+dependent on the action of its child completing before it can start. Finally, 
+the planner/cost-based-optimizer (CBO) will run various updates on the IR to
```
Comment (mosabua) [inline _episodes/11.md:76]: runs various updates

File: _episodes/11.md [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+optimize the query plan until it is ready to be executed. To sum it all up, the 
+planner and CBO generate and optimize the plan by running optimization rules. 
+Refer to chapter 4 in 
+[Presto: The Definitive Guide](https://www.starburstdata.com/oreilly-presto-guide-download/)
+pg. 50 for more information.
+
+In episode 9, we discussed how hash-joins work by first drawing a nested-loop
```
Comment (mosabua) [inline _episodes/11.md:83]: episode nine 

maybe

File: _episodes/11.md [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+analogy to how joins work. We then discussed how it is advantageous to read the
+inner loop into memory to avoid a lot of extra disk calls. Since it is ideal to
+read an entire table into memory, you will likely want to make sure the table
```
Comment (mosabua) [inline _episodes/11.md:86]: you likely want to make sure

File: _episodes/11.md [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+that is built in memory is the smaller size of the two tables. This smaller
+table called the build table. The table that gets streamed is called the probe
+table. We discussed a bit how hash-joins work which is a common mechanism to
+execute joins in a distributed and parallel fashion.  
+
+Another nomenclature akin to build table and probe tables are dimension and
+fact table, respectively. This nomenclature comes from the [star schema](https://en.wikipedia.org/wiki/Star_schema)
+that comes from data warehousing. Typically, there are large tables called fact
```
Comment (mosabua) [inline _episodes/11.md:94]: from the star schema tables from data warehousing.

File: _episodes/11.md [hunk 24]
```diff
@@ -0,0 +1,247 @@
...
+tables that would live at the center of the schema. These tables typically have
+many foreign keys, and a bit of quantitative or measuarable columns of the event
+or instance. The foreign keys connect these big fact tables to smaller dimension
+tables that when joined provide human readable context to enrich the recordings
```
Comment (mosabua) [inline _episodes/11.md:98]: that, when joined, provide...

File: _episodes/11.md [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+in the fact table. The schema ends up looking like a star with the fact table at
+the center. In essence, you just need to remember when someone is describing a
+fact table they are saying it is a bigger table that is likely going to end up
+on the probe side of a join, where a dimension is more likely a candidate to fit
+into memory on the build side of a join.
+
+So let's get onto the Dynamic Filtering shall we? First,let's cover a few
```
Comment (mosabua) [inline _episodes/11.md:105]: dynamic filtering

File: _episodes/11.md [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+concepts about Dynamic Filtering, then cover some variations of this concept.
```
Comment (mosabua) [inline _episodes/11.md:106]: dynamic filtering

File: _episodes/11.md [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+
+Dynamic Filtering takes advantage of joins with big fact tables to smaller
```
Comment (mosabua) [inline _episodes/11.md:108]: filtering

File: _episodes/11.md [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+dimension tables. What makes this filtering different from other types of
+filtering is that you are using the smaller build table that is loaded at query
+time to generate a list of values that exist in the join column between the
+build table and probe table. We know that only values that match these criteria
+are going to be returned from the sales side, so we can use this dynamically
+generated list as a pushdown predicate on the join column of the probe side.
+This means we are still scanning this data but only sending the subset that will
```
Comment (mosabua) [inline _episodes/11.md:115]: data, but only sending..

File: _episodes/11.md [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
+answer the query. We can look at the blog written for the original local dynamic
+filtering implementation by Roman Zeyde for more insights on the original
+implementation for Dynamic Filtering before Raunaq's changes.
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+Local Dynamic Filtering is definitely beneficial as it allows skipping
```
Comment (mosabua) [inline _episodes/11.md:122]: Local dynamic filtering

File: _episodes/11.md [hunk 30]
```diff
@@ -0,0 +1,247 @@
...
+generated list as a pushdown predicate on the join column of the probe side.
+This means we are still scanning this data but only sending the subset that will
+answer the query. We can look at the blog written for the original local dynamic
+filtering implementation by Roman Zeyde for more insights on the original
+implementation for Dynamic Filtering before Raunaq's changes.
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+Local Dynamic Filtering is definitely beneficial as it allows skipping 
+unnecessary stripes or row-groups in the orc or parquet reader. However, it
```
Comment (mosabua) [inline _episodes/11.md:123]: ORC or Parquet reader

File: _episodes/11.md [hunk 31]
```diff
@@ -0,0 +1,247 @@
...
+works only for broadcast joins and it's effectiveness depends upon the
```
Comment (mosabua) [inline _episodes/11.md:124]: joins, and it's effectiveness..

File: _episodes/11.md [hunk 32]
```diff
@@ -0,0 +1,247 @@
...
+implementation for Dynamic Filtering before Raunaq's changes.
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+Local Dynamic Filtering is definitely beneficial as it allows skipping 
+unnecessary stripes or row-groups in the orc or parquet reader. However, it
+works only for broadcast joins and it's effectiveness depends upon the 
+selectivity of the min and max indices maintained in orc or parquet files. What
+if we could prune entire partitions from the query execution based on dynamic
+filters? In the next iteration of Dynamic Filtering, called Dynamic Partition
```
Comment (mosabua) [inline _episodes/11.md:127]: dynamic filtering

File: _episodes/11.md [hunk 33]
```diff
@@ -0,0 +1,247 @@
...
+
+<https://trino.io/blog/2019/06/30/dynamic-filtering.html>
+
+Local Dynamic Filtering is definitely beneficial as it allows skipping 
+unnecessary stripes or row-groups in the orc or parquet reader. However, it
+works only for broadcast joins and it's effectiveness depends upon the 
+selectivity of the min and max indices maintained in orc or parquet files. What
+if we could prune entire partitions from the query execution based on dynamic
+filters? In the next iteration of Dynamic Filtering, called Dynamic Partition 
+Pruning, we do just that. We take advantage of the partitioned layout of Hive
```
Comment (mosabua) [inline _episodes/11.md:128]: dynamic partition pruning

File: _episodes/11.md [hunk 34]
```diff
@@ -0,0 +1,247 @@
...
+if we could prune entire partitions from the query execution based on dynamic
+filters? In the next iteration of Dynamic Filtering, called Dynamic Partition 
+Pruning, we do just that. We take advantage of the partitioned layout of Hive 
+tables to avoid generating splits on partitions that won't exist in the final
+query result. The coordinator can identify partitions for pruning based on the 
+dynamic filters sent to it from the workers processing the build side of join.
+This only works if the query contains a join condition on a column that is 
+partitioned.
+
+With that basic understanding, let's move on to the PR that implemented Dynamic
```
Comment (mosabua) [inline _episodes/11.md:135]: dynamic partition pruning

File: _episodes/11.md [hunk 35]
```diff
@@ -0,0 +1,247 @@
...
+Partition Pruning!
+
+## PR of the week: PR 1072 Implement Dynamic Partition Pruning
```
Comment (mosabua) [inline _episodes/11.md:138]: Implement dynamic partition pruning

File: _episodes/11.md [hunk 36]
```diff
@@ -0,0 +1,247 @@
...
+
+In this week's pull request <https://github.com/trinodb/trino/pull/1072> we
+return with Raunaq Morarka and Karol Sobczak. This PR effectively brings in the 
+second iteration of Dynamic Filtering, dynamic partition pruning, where instead
```
Comment (mosabua) [inline _episodes/11.md:142]: dynamic filtering - dynamic partition pruning.

Now start a new sentence.

File: _episodes/11.md [hunk 37]
```diff
@@ -0,0 +1,247 @@
...
+of relying on local dynamic filtering we collect dynamic filters from the
+workers in the coordinator and prune out extra splits that aren't needed with
+the partition layout of the probe side table. A query like this for example, 
+seen in Raunaq's blog about Dynamic Partition Pruning
```
Comment (mosabua) [inline _episodes/11.md:146]: like so

```
[Raunaq's blog about dynamic partition pruning]()
```

Comment (mosabua) [reply _episodes/11.md:146]: and in bracket us the post link syntax

File: _episodes/11.md [hunk 38]
```diff
@@ -0,0 +1,247 @@
...
+</p>
+<p align="center">
+6:
+<img align="center" width="50%" height="100%" src="/assets/episode/11/Dynamic
+ Filtering6.png"/>
+</p>
+
+## PR Demo: PR 1072 Implement Dynamic Partition Pruning
+
+For this PR demo, we have set up 1 r5.4xlarge coordinator and 4 r5.4xlarge
```
Comment (mosabua) [inline _episodes/11.md:194]: set up one r.54xlarge coordinator and four ...

File: _episodes/11.md [hunk 39]
```diff
@@ -0,0 +1,247 @@
...
+workers in a cluster. We have a sf100 size tpcds dataset. We will run some of
+the TPC-DS queries and perhaps a few others.
+
+The first query we run through in the TPC-DS queries was [query 54](https://github.com/trinodb/trino/blob/master/testing/trino-product-tests/src/main/resources/sql-tests/testcases/tpcds/q54.sql).
+With this query, we are using the `hive` catalog pointing to AWS S3 and AWS Glue
+as our metastore. We initially disable dynamic filtering then compare it to
```
Comment (mosabua) [inline _episodes/11.md:200]: filtering, then compare

File: _episodes/11.md [hunk 40]
```diff
@@ -0,0 +1,247 @@
...
+the times when dynamic filtering is enabled. With dynamic filtering we find the
+query to run at about 92 seconds, where with dynamic filtering it runs for 42
+seconds. We see similar findings for the semijoin we execute below and discuss
+some of the implications of how the planner actually optimizes the semijoin into
+an inner join.
+
+{% highlight SQL %}
```
Comment (mosabua) [inline _episodes/11.md:207]: use normal non-liquid code snippet syntax ideally

File: _episodes/11.md [hunk 41]
```diff
@@ -0,0 +1,247 @@
...
+SELECT ss_sold_date_sk, COUNT(*) from store_sales WHERE ss_sold_date_sk IN (
+  SELECT ws_sold_date_sk FROM (
+    SELECT ws_sold_date_sk, COUNT(*) FROM web_sales GROUP BY 1 ORDER BY 2 LIMIT 100
+  )
+)
+GROUP BY 1
+{% endhighlight %}
+
+
+## Events, news, and various links
```
Comment (mosabua) [inline _episodes/11.md:221]: empty line after title

**Review by mosabua — CHANGES_REQUESTED** [review]
Also update all the file names to non-space, lowercase please.

**Review by mosabua — APPROVED** [review]
Thanks for addressing everything! Good to go!

**Review by sopel39 — APPROVED** [review]
